<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Accessing and manipulating biological sequences –</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-dc64d301017c2e249433d9b3d1a3f4c9.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="https://www2.hshsl.umaryland.edu/cdabs/" class="navbar-brand navbar-brand-logo">
    <img src="https://www2.hshsl.umaryland.edu/hshsl/bin/img/logo_white@2x.png" alt="HSHSL logo" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="https://www2.hshsl.umaryland.edu/cdabs/">
    <span class="navbar-title"></span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./01-GettingSetUp.html"> 
<span class="menu-text">Getting Set Up</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./02-rentrez.html" aria-current="page"> 
<span class="menu-text">Finding and Manipulating Sequences</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./03-msa.html"> 
<span class="menu-text">Building a Bioinformatics Workflow</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#ncbi-data-and-the-rentrez-package" id="toc-ncbi-data-and-the-rentrez-package" class="nav-link" data-scroll-target="#ncbi-data-and-the-rentrez-package">NCBI data and the <strong><code>rentrez</code></strong> package</a></li>
  <li><a href="#finding-and-fetching-a-sequence" id="toc-finding-and-fetching-a-sequence" class="nav-link" data-scroll-target="#finding-and-fetching-a-sequence">Finding and fetching a sequence</a>
  <ul class="collapse">
  <li><a href="#try-it-yourself" id="toc-try-it-yourself" class="nav-link" data-scroll-target="#try-it-yourself">Try it Yourself!</a></li>
  </ul></li>
  <li><a href="#manipulating-sequences-with-biostrings" id="toc-manipulating-sequences-with-biostrings" class="nav-link" data-scroll-target="#manipulating-sequences-with-biostrings">Manipulating sequences with <strong><code>Biostrings</code></strong></a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://hshsl-training.github.io/r-for-bioinformatics/edit/main/02-rentrez.qmd" class="toc-action"><i class="bi bi-git"></i>Edit this page</a></li><li><a href="https://hshsl-training.github.io/r-for-bioinformatics/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Accessing and manipulating biological sequences</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<ul>
<li>Be able to name some main NCBI databases and their purposes</li>
<li>Become familiar with the <code>rentrez</code> core functions and how they are related.</li>
<li>Understand how to programmatically access NCBI data.</li>
<li>Use R to analyze and manipulate a biological sequence</li>
</ul>
</section>
<section id="ncbi-data-and-the-rentrez-package" class="level2">
<h2 class="anchored" data-anchor-id="ncbi-data-and-the-rentrez-package">NCBI data and the <strong><code>rentrez</code></strong> package</h2>
<p>The <a href="https://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information</a> (NCBI), a division of the National Library of Medicine, maintains several databases that provide access to a wide range of biological data and tools, including databases for DNA and protein sequences (GenBank, RefSeq), scientific literature (PubMed, PMC), and resources for analyzing genetic and genomic information (BLAST, Gene, SRA toolkit). It serves as a central hub for researchers around the world to find, share, and analyze biological data. NCBI is a major partner in the International Nucleotide Sequence Database Collaboration (INSDC) along with EMBL-EBI in Europe and DDBJ in Japan.</p>
<p>Entrez is the name of the integrated search system across all NCBI databases. Programmatic access to the Entrez system is provided by the <a href="https://www.ncbi.nlm.nih.gov/books/NBK25501/">Entrez Programming Utilities</a> (E-utilities). The E-utilities can be used on the command line, but in this class, we will be using the <strong><code>rentrez</code></strong> package, an R interface to the program.</p>
<p>First, let’s see what databases we can access with <code>rentrez</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">entrez_dbs</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "pubmed"          "protein"         "nuccore"         "ipg"            
 [5] "nucleotide"      "structure"       "genome"          "annotinfo"      
 [9] "assembly"        "bioproject"      "biosample"       "blastdbinfo"    
[13] "books"           "cdd"             "clinvar"         "gap"            
[17] "gapplus"         "grasp"           "dbvar"           "gene"           
[21] "gds"             "geoprofiles"     "medgen"          "mesh"           
[25] "nlmcatalog"      "omim"            "orgtrack"        "pmc"            
[29] "proteinclusters" "pcassay"         "protfam"         "pccompound"     
[33] "pcsubstance"     "seqannot"        "snp"             "sra"            
[37] "taxonomy"        "biocollections"  "gtr"            </code></pre>
</div>
</div>
<p>We can also see what fields are available to search in a database</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">entrez_db_searchable</span>(<span class="at">db=</span><span class="st">"gene"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Searchable fields for database 'gene'
  ALL    All terms from all searchable fields 
  UID    Unique number assigned to a gene record 
  FILT   Limits the records 
  TITL   gene or protein name 
  WORD   Free text associated with record 
  ORGN   scientific and common names of organism 
  MDAT   The last date on which the record was updated 
  CHR    Chromosome number or numbers; also 'mitochondrial', 'unknown' properties 
  MV     Chromosomal map location as displayed in MapViewer 
  GENE   Symbol or symbols of the gene 
  ECNO   EC number for enzyme or CAS registry number 
  MIM    MIM number from OMIM 
  DIS    Name(s) of diseases associated with this gene. When available, OMIM name will be used 
  ACCN   Nucleotide or protein accession(s) associated with this gene 
  UGEN   UniGene cluster number for this gene 
  PROP   Properties of Gene record 
  CDAT   The date on which this record first appeared 
  NCAC   nucleotide accessions of sequences 
  NUID   nucleotide uids of sequences 
  PACC   protein accessions 
  PUID   protein uids 
  PMID   PubMed ids of accessions linked to the record 
  TID    taxonomy id 
  GO     Gene Ontology 
  DOM    Domain Name 
  DDAT   The date on which the record was discontinued 
  CPOS   Chromosome base position 
  GFN    Gene full name 
  PFN    Protein full name 
  GL     Gene length 
  XC     Exon count 
  GRP    Relationships for this gene 
  PREF   Preferred symbol of the gene 
  AACC   Assembly accession 
  ASM    Assembly name 
  EXPR   Gene expression </code></pre>
</div>
</div>
<p>The functions of <code>rentrez</code> are designed to work together. The main ones we will be using in this session are:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 27%">
<col style="width: 73%">
</colgroup>
<thead>
<tr class="header">
<th>Function</th>
<th>Purpose</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>entrez_search()</code></td>
<td>Search a specific NCBI database and retrieve UIDs (unique identifiers).</td>
</tr>
<tr class="even">
<td><code>entrez_summary()</code></td>
<td>Get summary metadata for UIDs returned by a search.</td>
</tr>
<tr class="odd">
<td><code>extract_from_esummary()</code></td>
<td>Extract specific fields (e.g., title, publication date) from esummary results.</td>
</tr>
<tr class="even">
<td><code>entrez_fetch()</code></td>
<td>Download full records (e.g., FASTA, GenBank, XML).</td>
</tr>
<tr class="odd">
<td><code>entrez_link()</code></td>
<td>Find related data across NCBI databases.</td>
</tr>
</tbody>
</table>
<p>You can check out the <a href="https://docs.ropensci.org/rentrez/"><code>rentrez</code> documentation</a> for more information on these and other functions.</p>
</section>
<section id="finding-and-fetching-a-sequence" class="level2">
<h2 class="anchored" data-anchor-id="finding-and-fetching-a-sequence">Finding and fetching a sequence</h2>
<p>Let’s start by searching for the record in the Gene database for human gene TP53, an important gene in the study of cancer. The <code>entrez_search()</code> function takes the following arguments:</p>
<ul>
<li><code>db</code>, name of the database to search for (required)</li>
<li><code>term</code>, the search term, you can also use MeSH terms to perform your search (required)</li>
<li><code>retmax</code>, the default of retrievable ids is 20, this argument can be used to change that number</li>
<li><code>retmode</code>, to select the format of your output (XML or JSON), by default will be XML</li>
<li><code>use_history</code>, to store a history of searches in NCBI’s server</li>
</ul>
<p>The <code>term</code> argument can take a simple or more complicated search string linked together with boolean terms <strong>AND, OR, and NOT</strong>. The name of the field like <code>[gene name]</code> or <code>[orgn]</code> comes after the search term for that field. Note that the <code>[orgn]</code> field can accept scientific or common names or organisms.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>search_tp53_hs <span class="ot">&lt;-</span> <span class="fu">entrez_search</span>(<span class="at">db=</span><span class="st">"gene"</span>, </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">term =</span><span class="st">"TP53[gene] AND human[orgn]"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Lists
</div>
</div>
<div class="callout-body-container callout-body">
<p>Let’s take a closer look at the results we get from our search. The <code>search_tp53_hs</code> object is a structure called a <strong>list</strong>. Recall that a vector is a one-dimensional data structure in R where every element must be of the same type. Lists are one-dimensional data structures where elements can be anything, even data frames or other lists. Lists are a common way to store hierarchical data, like XML or JSON data formats. You’ll find that it is very common for data to be stored as a list in bioinformatics workflows.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 14%">
<col style="width: 42%">
<col style="width: 43%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Homogeneous</strong></th>
<th><strong>Heterogeneous</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>1D</strong></td>
<td><strong>Vector</strong><br>All elements of the same type</td>
<td><strong>List</strong><br>Can contain any type of R object</td>
</tr>
<tr class="even">
<td><strong>2D</strong></td>
<td><strong>Matrix</strong><br>All elements of the same type</td>
<td><strong>Data Frame</strong><br>Each column can differ in type</td>
</tr>
</tbody>
</table>
<p>To access items in a list, you can use <code>$</code> or <code>[[]]</code>, similar to accessing vectors in a dataframe. There are five elements in our search object:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 19%">
<col style="width: 80%">
</colgroup>
<thead>
<tr class="header">
<th><strong>Element</strong></th>
<th><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>count</code></td>
<td>Total number of records in the database that match the search query.</td>
</tr>
<tr class="even">
<td><code>retmax</code></td>
<td>The number of IDs returned in the <code>ids</code> vector (default is 20, can be changed).</td>
</tr>
<tr class="odd">
<td><code>ids</code></td>
<td>A character vector of NCBI record IDs that matched the search.</td>
</tr>
<tr class="even">
<td><code>QueryTranslation</code></td>
<td>Shows how NCBI interpreted and expanded your search query using synonyms etc.</td>
</tr>
<tr class="odd">
<td><code>file</code></td>
<td>A temporary file storing your search results (used internally by rentrez).</td>
</tr>
</tbody>
</table>
<p>We can look closer at the <code>ids</code> element:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>search_tp53_hs<span class="sc">$</span>ids</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "7157"</code></pre>
</div>
</div>
<p>or</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>search_tp53_hs[[<span class="dv">1</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "7157"</code></pre>
</div>
</div>
</div>
</div>
<p>The ids will likely be the most important part of the search object. You can use those ids with <code>entrez_summary()</code>, <code>entrez_link()</code> and <code>entrez_fetch()</code>.</p>
<p>Let’s use the id to see if there are records for human TP53 in the <strong>nucleotide</strong> database. While Gene contains metadata about genes, <strong>nucleotide</strong> contains the actual sequences. There are three main arguments you need to supply to the link function:</p>
<ul>
<li><code>dbfrom</code> character name of database from which the id(s) originate</li>
<li><code>ids</code> vector with unique ID(s) for records in database db</li>
<li><code>db</code> character name of database in which to search for links. Use “all” to search for links in all databases.</li>
</ul>
<p>So we are going to search from gene with the id we retrieved from our search, and look in “nuccore”, the name for <strong>nucleotide</strong> in the Entrez system.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>link_tp53_hs <span class="ot">&lt;-</span> <span class="fu">entrez_link</span>(<span class="at">dbfrom=</span><span class="st">"gene"</span>, </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">id=</span>search_tp53_hs<span class="sc">$</span>ids, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">db=</span><span class="st">"nuccore"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s take a closer look at the <code>links</code> element of our link object</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>link_tp53_hs<span class="sc">$</span>links</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>elink result with information from 4 databases:
[1] gene_nuccore            gene_nuccore_mgc        gene_nuccore_refseqgene
[4] gene_nuccore_refseqrna </code></pre>
</div>
</div>
<p>This is saying there are related records for this gene in the general nucleotide collection, as well as in three subsets: Mammalian Gene Collection, RefSeqGene and RefSeqRNA. We will be especially interested in the RefSeqGene sequence. RefSeq, short for Reference Sequence, is NCBI’s curated collection of consensus sequences. These are likely to be higher quality than other sequences, so it is a good idea to use them for your analysis when possible.</p>
<p>We can access the RefSeqGene id:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>link_tp53_hs<span class="sc">$</span>links<span class="sc">$</span>gene_nuccore_refseqgene</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "383209646"</code></pre>
</div>
</div>
<p>Now we have the id for the sequence we want to look at, we use <code>entrez_fetch()</code>, which requires three arguments</p>
<ul>
<li><code>db</code> character name of the database to use</li>
<li><code>id</code> vector of unique ID(s) for records in database <code>db</code></li>
<li><code>rettype</code> - character name of the format in which to return record, such as fasta or xml. We will ask for our record in fasta format, a common plain text way of representing sequence data.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>fetch_tp53_hs <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db=</span><span class="st">"nuccore"</span>,</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">id=</span>link_tp53_hs<span class="sc">$</span>links<span class="sc">$</span>gene_nuccore_refseqgene, </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">rettype =</span> <span class="st">"fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s take a look at our sequence in the console. You can see it’s pretty simple, just a header line starting with <code>&gt;</code> and the sequence of bases.</p>
<p>When we print it to the console as usual, we see a lot of <code>\n</code> characters which is the newline symbol.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>fetch_tp53_hs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can use the cat function to print it out a little more nicely.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(fetch_tp53_hs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also save this to a file</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">writeLines</span>(fetch_tp53_hs, <span class="st">"data/tp53_human.fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="try-it-yourself" class="level3">
<h3 class="anchored" data-anchor-id="try-it-yourself">Try it Yourself!</h3>
<div class="callout callout-style-default callout-tip callout-titled" title="CHALLENGE">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
CHALLENGE
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Look up which fields are searchable in the <code>Protein</code> database.</li>
<li>Search <code>Gene</code> for the BRCA1 gene in humans. What id is returned?</li>
<li>Use <code>entrez_link()</code> to find <strong>all</strong> databases that have a link to this Gene record. How many databases are linked? How many related records are there in Pubmed?</li>
</ol>
<div class="callout callout-style-default callout-tip no-icon callout-titled" title="Solution">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li><code>entrez_db_searchable("protein")</code></li>
<li><code>brca1_search &lt;- entrez_search(db = "gene", term = "BRCA1[gene] AND human[orgn]")</code> <br> <code>brca1_search$ids</code></li>
<li><code>brca1_link &lt;- entrez_link(dbfrom = "gene", id = test_search$ids, db = "all")</code><br> <code>brca1_link$links</code><br> <code>brca1_links$links$gene_pubmed</code></li>
</ol>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="manipulating-sequences-with-biostrings" class="level2">
<h2 class="anchored" data-anchor-id="manipulating-sequences-with-biostrings">Manipulating sequences with <strong><code>Biostrings</code></strong></h2>
<p>Now we will use the <strong><code>Biostrings</code></strong> package from Bioconductor to read in our fasta file and save it as the specialized class <code>XStringSet</code> which essentially serves as a container sets of biological sequence data. The X can be replaced by the appropriate molecule: DNA, RNA, or AA for proteins. So here we are saving our object as a <code>DNAStringSet</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>fasta <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(<span class="st">"data/tp53_human.fasta"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>XStringSets</code> are made up of <code>XString</code> objects. That is, each sequence in a <code>XStringSet</code> is a <code>XString</code> object. Even though our fasta file only contains one sequence, we still save it as an <code>DNAStringSet</code>. We can specifically pull out our sequence into a <code>DNAStringSet</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>dna_tp53_human <span class="ot">&lt;-</span> fasta[[<span class="dv">1</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>As you work more with bioinformatics packages, you’ll find that many of them create these kinds of specialized classes, and you’ll be able to put together workflows of packages designed to work with each other’s data structures.</p>
<p>Now we have our <code>DNAString</code> object, we can use BioStrings to get summary statistics and manipulate it to other objects.</p>
<p>First we can use <code>alphabetFrequency</code> to get the frequence of bases in our sequence.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">alphabetFrequency</span>(dna_tp53_human)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   A    C    G    T    M    R    W    S    Y    K    V    H    D    B    N    - 
8647 7986 8058 8081    0    0    0    0    0    0    0    0    0    0    0    0 
   +    . 
   0    0 </code></pre>
</div>
</div>
<p>In addition to the four bases ACGT, it also contains several other codes that indicate ambiguous bases. This is a high quality consensus RefSeq sequence though, so we do not have any ambiguous bases.</p>
<p>We can specify we are only interested in ACGT counts by specifying the argument <code>baseOnly = TRUE</code></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">alphabetFrequency</span>(dna_tp53_human, <span class="at">baseOnly =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    A     C     G     T other 
 8647  7986  8058  8081     0 </code></pre>
</div>
</div>
<p>To look for a specific base or base, we can use <code>letterFrequency</code> instead of <code>alphabetFrequency</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>gc_content <span class="ot">&lt;-</span> <span class="fu">letterFrequency</span>(dna_tp53_human, </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">letters =</span> <span class="fu">c</span>(<span class="st">"G"</span>, <span class="st">"C"</span>), </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">as.prob=</span><span class="cn">TRUE</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(gc_content)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.4895643</code></pre>
</div>
</div>
<p>We can also use <code>Biostrings</code> to generate the complement of the DNA strand, the reverse, and the reverse complement.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>tp53_comp <span class="ot">&lt;-</span> <span class="fu">complement</span>(dna_tp53_human)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>tp53_rev <span class="ot">&lt;-</span> <span class="fu">reverse</span>(dna_tp53_human)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>tp53_rev_comp <span class="ot">&lt;-</span> <span class="fu">reverseComplement</span>(dna_tp53_human)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s compare:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>dna_tp53_human</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>tp53_comp</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>tp53_rev</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>tp53_rev_comp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>32772-letter DNAString object
seq: CTCCTTGGTTCAAGTAATTCTCCTGCCTCAGACTCC...AAAAAAAAGATACTACAAAGTCAAGAGACAAACAAT
32772-letter DNAString object
seq: GAGGAACCAAGTTCATTAAGAGGACGGAGTCTGAGG...TTTTTTTTCTATGATGTTTCAGTTCTCTGTTTGTTA
32772-letter DNAString object
seq: TAACAAACAGAGAACTGAAACATCATAGAAAAAAAA...CCTCAGACTCCGTCCTCTTAATGAACTTGGTTCCTC
32772-letter DNAString object
seq: ATTGTTTGTCTCTTGACTTTGTAGTATCTTTTTTTT...GGAGTCTGAGGCAGGAGAATTACTTGAACCAAGGAG</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/hshsl-training\.github\.io\/r-for-bioinformatics\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://hshsl-training.github.io/r-for-bioinformatics/edit/main/02-rentrez.qmd" class="toc-action"><i class="bi bi-git"></i>Edit this page</a></li><li><a href="https://hshsl-training.github.io/r-for-bioinformatics/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p><a href="https://creativecommons.org/licenses/by-sa/4.0/"><img src="https://mirrors.creativecommons.org/presskit/buttons/88x31/svg/by.svg" class="img-fluid" alt="CC BY" width="125"></a></p>
</div>
  </div>
</footer>




</body></html>